File: leetcode/2975. Maximum Square Area by Removing Fences From a Field/solution.cpp
#include <vector>
#include <algorithm>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int maximizeSquareArea(int m, int n, vector<int>& hFences, vector<int>& vFences) {
        long long MOD = 1e9 + 7;

        hFences.push_back(1);
        hFences.push_back(m);
        vFences.push_back(1);
        vFences.push_back(n);

        sort(hFences.begin(), hFences.end());
        sort(vFences.begin(), vFences.end());

        unordered_set<int> hGaps;
        int N = hFences.size();
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < i; ++j) {
                hGaps.insert(hFences[i] - hFences[j]);
            }
        }

        long long edgeMax = -1;
        int M = vFences.size();

        for (int i = 0; i < M; ++i) {
            for (int j = 0; j < i; ++j) {
                int gap = vFences[i] - vFences[j];
                if (hGaps.count(gap)) {
                    edgeMax = max(edgeMax, (long long)gap);
                }
            }
        }

        if (edgeMax == -1) {
            return -1;
        }

        return (edgeMax * edgeMax) % MOD;
    }
};

File: leetcode/2975. Maximum Square Area by Removing Fences From a Field/solution.go
func maximizeSquareArea(m int, n int, hFences []int, vFences []int) int {
    const MOD = 1e9 + 7
    hFences = append(hFences, 1)
    hFences = append(hFences, m)
    vFences = append(vFences, 1)
    vFences = append(vFences, n)

    sort.Slice(hFences, func (i, j int) bool {
        return hFences[i] < hFences[j]
    })

    sort.Slice(vFences, func (i, j int) bool {
        return vFences[i] < vFences[j]
    })

    N, M := len(hFences), len(vFences)

    mp := make(map[int]bool)

    for i := 0; i < N; i++ {
        for j := 0; j < i; j++ {
            mp[hFences[i] - hFences[j]] = true
        }
    }

    edgeMax := -1

    for i := 0; i < M; i++ {
        for j := 0; j < i; j++ {
            if mp[vFences[i] - vFences[j]] {
                edgeMax = max(edgeMax, vFences[i] - vFences[j])
            }
        }
    }

    if edgeMax == -1 {
        return -1
    }

    return (edgeMax % MOD)  * (edgeMax % MOD) % MOD
}

func max(a, b int) int {
    if a > b {
        return a
    }

    return b
}